api/v1alpha2/conversion.go

```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
 func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *v1alpha3.MachineSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Bootstrap.ConfigRef.APIVersion)
+	if err != nil {
+		return err
+	}
+
 	if err := autoConvert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in, out, s); err != nil {
 		return err
 	}
 
+	if inGV.Group == bootstrapv1a3.GroupVersion.Group && (out.Bootstrap.ConfigRef.Kind == kubeadmConfigKind || out.Bootstrap.ConfigRef.Kind == kubeadmConfigTemplateKind) {
+		out.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
+	}
```

Comment:
> NB: This comment applies to all conversions.

This and the check above can be simplified:

```golang
if in.Bootstrap.ConfigRef != nil {
		gvk := in.Bootstrap.ConfigRef.GetObjectKind().GroupVersionKind()
		if gvk.Group == bootstrapv1a3.GroupVersion.Group && gvk.Kind == kubeadmConfigTemplateKind {
			out.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
		}
	}
```

The template is the only reference we should change for MachineSet or MachineDeployment, while for Machines we only need to check the `KubeadmConfig` kind.

Reply:
> The template is the only reference we should change for MachineSet or MachineDeployment, while for Machines we only need to check the `KubeadmConfig` kind.

Based on this, now
-  for `MachineSpec`, checking only for `kubeadmConfigKind`
- for `MachineSetSpec` and `MachineDeploymentSpec`, checking only for `kubeadmConfigTemplateKind`

Also, based on my earlier understanding, I added tests for the conditions when a `MachineSet` object has `kubeadmConfigTemplate` kind reference and `MachineSetSpec` and `MachineDeploymentSpec` objects have `kubeadmConfig` kind references. For example- https://github.com/kubernetes-sigs/cluster-api/pull/1938/files#diff-c8f1f9d0a7bf481d8e912ebc493d3c9fR337